File: Ethereum/SmartContract/deploy.py
#!/usr/bin/env python3
"""
Python deployment script for MyERC20Token
Deploys the ERC20 token to the local Ethereum network
"""

import json
import os
import sys
from web3 import Web3
from eth_account import Account
import time

# Network configuration
RPC_URL = "http://127.0.0.1:32973"
CHAIN_ID = 585858

# Private keys (from the network's funded accounts)
PRIVATE_KEYS = [
    "27515f805127bebad2fb9b183508bdacb8c763da16f54e0678b16e8f28ef3fff",
    "7ff1a4c1d57e5e784d327c4c7651e952350bc271f156afb3d00d20f5ef924856",
    "3a91003acaf4c21b3953d94fa4a6db694fa69e5242b2e37be05dd82761058899"
]

# Contract ABI (simplified for ERC20)
ERC20_ABI = [
    {
        "inputs": [{"internalType": "address", "name": "recipient", "type": "address"}],
        "stateMutability": "nonpayable",
        "type": "constructor"
    },
    {
        "anonymous": False,
        "inputs": [
            {"indexed": True, "internalType": "address", "name": "owner", "type": "address"},
            {"indexed": True, "internalType": "address", "name": "spender", "type": "address"},
            {"indexed": False, "internalType": "uint256", "name": "value", "type": "uint256"}
        ],
        "name": "Approval",
        "type": "event"
    },
    {
        "anonymous": False,
        "inputs": [
            {"indexed": True, "internalType": "address", "name": "from", "type": "address"},
            {"indexed": True, "internalType": "address", "name": "to", "type": "address"},
            {"indexed": False, "internalType": "uint256", "name": "value", "type": "uint256"}
        ],
        "name": "Transfer",
        "type": "event"
    },
    {
        "inputs": [{"internalType": "address", "name": "owner", "type": "address"}, {"internalType": "address", "name": "spender", "type": "address"}],
        "name": "allowance",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [{"internalType": "address", "name": "spender", "type": "address"}, {"internalType": "uint256", "name": "amount", "type": "uint256"}],
        "name": "approve",
        "outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
        "stateMutability": "nonpayable",
        "type": "function"
    },
    {
        "inputs": [{"internalType": "address", "name": "account", "type": "address"}],
        "name": "balanceOf",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [],
        "name": "decimals",
        "outputs": [{"internalType": "uint8", "name": "", "type": "uint8"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [],
        "name": "name",
        "outputs": [{"internalType": "string", "name": "", "type": "string"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [],
        "name": "symbol",
        "outputs": [{"internalType": "string", "name": "", "type": "string"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [],
        "name": "totalSupply",
        "outputs": [{"internalType": "uint256", "name": "", "type": "uint256"}],
        "stateMutability": "view",
        "type": "function"
    },
    {
        "inputs": [{"internalType": "address", "name": "to", "type": "address"}, {"internalType": "uint256", "name": "amount", "type": "uint256"}],
        "name": "transfer",
        "outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
        "stateMutability": "nonpayable",
        "type": "function"
    },
    {
        "inputs": [{"internalType": "address", "name": "from", "type": "address"}, {"internalType": "address", "name": "to", "type": "address"}, {"internalType": "uint256", "name": "amount", "type": "uint256"}],
        "name": "transferFrom",
        "outputs": [{"internalType": "bool", "name": "", "type": "bool"}],
        "stateMutability": "nonpayable",
        "type": "function"
    }
]

def connect_to_network():
    """Connect to the Ethereum network"""
    print("🔗 Connecting to Ethereum network...")
    w3 = Web3(Web3.HTTPProvider(RPC_URL))
    
    if not w3.is_connected():
        print("❌ Failed to connect to network")
        return None
    
    print(f"✅ Connected to network at {RPC_URL}")
    return w3

def get_account(w3, private_key):
    """Get account from private key"""
    account = Account.from_key(private_key)
    return account

def check_balance(w3, address):
    """Check ETH balance"""
    balance = w3.eth.get_balance(address)
    return w3.from_wei(balance, 'ether')

def deploy_contract(w3, account, contract_bytecode, constructor_args):
    """Deploy the contract"""
    print("📦 Deploying MyERC20Token contract...")
    
    # Get nonce
    nonce = w3.eth.get_transaction_count(account.address)
    
    # Build transaction
    transaction = {
        'from': account.address,
        'gas': 2000000,  # Gas limit
        'gasPrice': w3.to_wei('20', 'gwei'),
        'nonce': nonce,
        'data': contract_bytecode + constructor_args.hex()[2:] if constructor_args else contract_bytecode
    }
    
    # Sign transaction
    signed_txn = w3.eth.account.sign_transaction(transaction, account.key)
    
    # Send transaction
    tx_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
    print(f"📝 Transaction sent: {tx_hash.hex()}")
    
    # Wait for receipt
    print("⏳ Waiting for transaction confirmation...")
    receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
    
    if receipt.status == 1:
        print("✅ Contract deployed successfully!")
        return receipt.contractAddress
    else:
        print("❌ Contract deployment failed")
        return None

def get_contract_info(w3, contract_address, account):
    """Get contract information"""
    print("\n📊 Contract Information:")
    
    # Create contract instance
    contract = w3.eth.contract(address=contract_address, abi=ERC20_ABI)
    
    try:
        name = contract.functions.name().call()
        symbol = contract.functions.symbol().call()
        decimals = contract.functions.decimals().call()
        total_supply = contract.functions.totalSupply().call()
        balance = contract.functions.balanceOf(account.address).call()
        
        print(f"   Name: {name}")
        print(f"   Symbol: {symbol}")
        print(f"   Decimals: {decimals}")
        print(f"   Total Supply: {total_supply / (10 ** decimals)} {symbol}")
        print(f"   Your Balance: {balance / (10 ** decimals)} {symbol}")
        
        return {
            'name': name,
            'symbol': symbol,
            'decimals': decimals,
            'totalSupply': total_supply,
            'balance': balance
        }
    except Exception as e:
        print(f"❌ Error getting contract info: {e}")
        return None

def save_deployment_info(contract_address, deployer, tx_hash, contract_info):
    """Save deployment information"""
    deployment_info = {
        'network': 'localhost',
        'contractAddress': contract_address,
        'deployer': deployer,
        'transactionHash': tx_hash,
        'timestamp': time.strftime('%Y-%m-%d %H:%M:%S'),
        'contractInfo': contract_info,
        'rpcUrl': RPC_URL,
        'chainId': CHAIN_ID
    }
    
    # Create deployments directory
    os.makedirs('deployments', exist_ok=True)
    
    # Save to file
    filename = f'deployments/deployment-{int(time.time())}.json'
    with open(filename, 'w') as f:
        json.dump(deployment_info, f, indent=2)
    
    print(f"💾 Deployment info saved to: {filename}")
    return filename

def main():
    print("🚀 MyERC20Token Python Deployment Script")
    print("=" * 50)
    
    # Connect to network
    w3 = connect_to_network()
    if not w3:
        return
    
    # Get account
    account = get_account(w3, PRIVATE_KEYS[0])
    print(f"📝 Using account: {account.address}")
    
    # Check balance
    balance = check_balance(w3, account.address)
    print(f"💰 Account balance: {balance} ETH")
    
    if balance == 0:
        print("⚠️  Warning: Account has no ETH. Make sure the network is running and funded.")
        return
    
    # Note: In a real deployment, you would need the compiled bytecode
    # For this example, we'll show the structure
    print("\n📝 Note: This script requires the compiled contract bytecode.")
    print("   To get the bytecode, compile the contract using Hardhat:")
    print("   npm install")
    print("   npx hardhat compile")
    print("   Then extract the bytecode from artifacts/contracts/MyERC20Token.sol/MyERC20Token.json")
    
    print("\n🌐 Network Information:")
    print(f"   RPC URL: {RPC_URL}")
    print(f"   Chain ID: {CHAIN_ID}")
    print("   Block Explorer (Dora): http://127.0.0.1:32826")
    print("   Block Explorer (Blockscout): http://127.0.0.1:3000")
    
    print("\n💡 For actual deployment, use the Hardhat script:")
    print("   npm install")
    print("   npx hardhat run scripts/deploy.js --network localhost")

if __name__ == "__main__":
    main()

File: Ethereum/SmartContract/deploy.js
const { Web3 } = require('web3');
const fs = require('fs');
const solc = require('solc');

// Configuration
const CONFIG = {
    RPC_URL: 'http://localhost:8545',
    CHAIN_ID: 1337,
    CONTRACT_FILE: './SimpleERC20Token.sol',
    CONTRACT_NAME: 'SimpleERC20Token',
    INITIAL_SUPPLY: 1000, // 1000 tokens
    GAS_PRICE: '20000000000', // 20 Gwei
    GAS_LIMIT: 3000000,
    // Development private key (DO NOT USE IN PRODUCTION!)
    PRIVATE_KEY: '0x1234567890123456789012345678901234567890123456789012345678901234'
};

// Colors for console output
const colors = {
    reset: '\x1b[0m',
    bright: '\x1b[1m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    magenta: '\x1b[35m',
    cyan: '\x1b[36m'
};

function log(message, color = 'reset') {
    console.log(`${colors[color]}${message}${colors.reset}`);
}

class ERC20TokenDeployer {
    constructor() {
        this.web3 = null;
        this.account = null;
        this.contract = null;
        this.contractAddress = null;
    }

    async initialize() {
        log('🚀 Initializing ERC20 Token Deployer...', 'cyan');
        try {
            // Initialize Web3
            this.web3 = new Web3(CONFIG.RPC_URL);
            
            // Check connection
            const isConnected = await this.web3.eth.net.isListening();
            if (!isConnected) {
                throw new Error('Cannot connect to Ethereum node. Make sure the network is running.');
            }
            
            // Get network info
            const networkId = await this.web3.eth.net.getId();
            if (parseInt(networkId) !== CONFIG.CHAIN_ID) {
                throw new Error(`Wrong network. Expected chain ID ${CONFIG.CHAIN_ID}, got ${networkId}`);
            }
            
            log(`✅ Connected to network (Chain ID: ${networkId})`, 'green');
            
            // Create account from private key
            this.account = this.web3.eth.accounts.privateKeyToAccount(CONFIG.PRIVATE_KEY);
            this.web3.eth.accounts.wallet.add(this.account);
            
            log(`✅ Account created from private key`, 'green');
            log(`📝 Deployer account: ${this.account.address}`, 'blue');
            
            // Check balance
            const balance = await this.web3.eth.getBalance(this.account.address);
            const balanceEth = this.web3.utils.fromWei(balance, 'ether');
            log(`💰 Account balance: ${balanceEth} ETH`, 'blue');
            
            if (balance === '0') {
                log(`⚠️ Account has no ETH. This account needs to be funded in the genesis block.`, 'yellow');
                log(`💡 For development, you can fund this account manually or use an existing funded account.`, 'yellow');
                throw new Error('Account has no ETH. Please fund this account first.');
            }

        } catch (error) {
            log(`❌ Initialization failed: ${error.message}`, 'red');
            throw error;
        }
    }

    async compileContract() {
        log('🔨 Compiling ERC20 smart contract...', 'yellow');
        try {
            const contractCode = fs.readFileSync(CONFIG.CONTRACT_FILE, 'utf8');
            
            // Create input for solc with OpenZeppelin dependencies
            const input = {
                language: 'Solidity',
                sources: {
                    'SimpleERC20Token.sol': {
                        content: contractCode,
                    },
                    '@openzeppelin/contracts/token/ERC20/ERC20.sol': {
                        content: fs.readFileSync('./node_modules/@openzeppelin/contracts/token/ERC20/ERC20.sol', 'utf8')
                    },
                    '@openzeppelin/contracts/token/ERC20/IERC20.sol': {
                        content: fs.readFileSync('./node_modules/@openzeppelin/contracts/token/ERC20/IERC20.sol', 'utf8')
                    },
                    '@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol': {
                        content: fs.readFileSync('./node_modules/@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol', 'utf8')
                    },
                    '@openzeppelin/contracts/utils/Context.sol': {
                        content: fs.readFileSync('./node_modules/@openzeppelin/contracts/utils/Context.sol', 'utf8')
                    }
                },
                settings: {
                    outputSelection: {
                        '*': {
                            '*': ['*'],
                        },
                    },
                },
            };

            const output = JSON.parse(solc.compile(JSON.stringify(input)));
            
            if (output.errors && output.errors.length > 0) {
                console.log('Compilation errors:', output.errors);
            }
            
            const contractOutput = output.contracts['SimpleERC20Token.sol']['SimpleERC20Token'];

            if (!contractOutput) {
                throw new Error(`Contract ${CONFIG.CONTRACT_NAME} not found in compilation output.`);
            }

            this.contractAbi = contractOutput.abi;
            this.contractBytecode = contractOutput.evm.bytecode.object;

            // Save compiled contract to build directory
            const buildDir = './build';
            if (!fs.existsSync(buildDir)) {
                fs.mkdirSync(buildDir);
            }
            fs.writeFileSync(`${buildDir}/SimpleERC20Token.json`, JSON.stringify(contractOutput, null, 2));
            log('✅ Contract compiled successfully', 'green');
            log('💾 Compiled contract saved to build/SimpleERC20Token.json', 'blue');

        } catch (error) {
            log(`❌ Compilation failed: ${error.message}`, 'red');
            throw error;
        }
    }

    async deployContract() {
        log('🚀 Deploying ERC20 smart contract...', 'cyan');
        try {
            this.contract = new this.web3.eth.Contract(this.contractAbi);
            
            // Prepare deployment data - the constructor takes a recipient address
            const deployData = this.contract.deploy({
                data: this.contractBytecode,
                arguments: [this.account.address] // Deployer receives the initial tokens
            });
            
            // Get nonce
            const nonce = await this.web3.eth.getTransactionCount(this.account.address);
            
            // Create transaction object
            const tx = {
                from: this.account.address,
                data: deployData.encodeABI(),
                gas: CONFIG.GAS_LIMIT,
                gasPrice: CONFIG.GAS_PRICE,
                nonce: nonce
            };
            
            log(`📤 Sending signed transaction...`, 'yellow');
            log(`🔢 Nonce: ${nonce}`, 'blue');
            log(`⛽ Gas: ${CONFIG.GAS_LIMIT}`, 'blue');
            log(`💰 Gas Price: ${CONFIG.GAS_PRICE}`, 'blue');
            log(`🎯 Initial recipient: ${this.account.address}`, 'blue');
            
            // Sign and send transaction
            const signedTx = await this.web3.eth.accounts.signTransaction(tx, CONFIG.PRIVATE_KEY);
            const receipt = await this.web3.eth.sendSignedTransaction(signedTx.rawTransaction);
            
            this.contractAddress = receipt.contractAddress;
            
            log(`✅ Contract deployed successfully!`, 'green');
            log(`📍 Contract address: ${this.contractAddress}`, 'green');
            log(`🔗 Transaction hash: ${receipt.transactionHash}`, 'green');

            // Save contract address and ABI
            fs.writeFileSync('./build/SimpleERC20Token.address', this.contractAddress);
            fs.writeFileSync('./build/SimpleERC20Token.abi', JSON.stringify(this.contractAbi, null, 2));
            log('📝 Contract address and ABI saved to build/ directory', 'green');

            // Test the deployed contract
            await this.testContract();

        } catch (error) {
            log(`❌ Deployment failed: ${error.message}`, 'red');
            log(`🔍 Error details: ${JSON.stringify(error, null, 2)}`, 'yellow');
            throw error;
        }
    }

    async testContract() {
        log('🧪 Testing deployed contract...', 'cyan');
        try {
            const contractInstance = new this.web3.eth.Contract(this.contractAbi, this.contractAddress);
            
            // Test token name
            const name = await contractInstance.methods.name().call();
            log(`📛 Token name: ${name}`, 'green');
            
            // Test token symbol
            const symbol = await contractInstance.methods.symbol().call();
            log(`🔤 Token symbol: ${symbol}`, 'green');
            
            // Test total supply
            const totalSupply = await contractInstance.methods.totalSupply().call();
            const totalSupplyFormatted = this.web3.utils.fromWei(totalSupply, 'ether');
            log(`📊 Total supply: ${totalSupplyFormatted} ${symbol}`, 'green');
            
            // Test balance of deployer
            const balance = await contractInstance.methods.balanceOf(this.account.address).call();
            const balanceFormatted = this.web3.utils.fromWei(balance, 'ether');
            log(`💰 Deployer balance: ${balanceFormatted} ${symbol}`, 'green');
            
            // Test decimals
            const decimals = await contractInstance.methods.decimals().call();
            log(`🔢 Decimals: ${decimals}`, 'green');

        } catch (error) {
            log(`⚠️ Contract test failed: ${error.message}`, 'yellow');
        }
    }

    async run() {
        try {
            await this.initialize();
            await this.compileContract();
            await this.deployContract();
            log('🎉 ERC20 Token deployment completed successfully!', 'green');
        } catch (error) {
            log(`\n💥 Deployment failed: ${error.message}`, 'red');
        }
    }
}

const deployer = new ERC20TokenDeployer();
deployer.run();